src/backend/env/tip.rs

```diff
@@ -0,0 +1,385 @@
+        .principal_to_user(to_principal)
+        .ok_or("receiver not found")?
+        .id;
+    let (sender_id, sender_name) = state
+        .principal_to_user(from_principal)
+        .map(|sender| (sender.id, sender.name.clone()))
+        .ok_or("sender not found")?;
+    let realm = post
+        .realm
+        .as_ref()
+        .and_then(|id| state.realms.get(id))
+        .ok_or("realm not found")?;
+    let memo_u64 = convert_memo_to_u64(memo)?;
+
+    if post.user != receiver_id {
+        return Err("receiver does not match with post creator".to_string());
+    }
+    if realm.native_token != Some(canister_id)
+        && realm
+            .tokens
+            .as_ref()
+            .map_or(true, |tokens| !tokens.contains(&canister_id))
+    {
+        return Err("token is not allowed to tip in realm".to_string());
+    }
+
+    if memo_u64 != post_id {
+        return Err(format!(
+            "memo {} does not match with post id {}",
+            memo_u64, post_id
+        ));
+    }
+
+    let has_external_tip = post.has_external_tip;
+
+    if state
+        .post_tip_indexes
+        .get(&post_id)
+        .map_or(false, |tip_ids| {
+            tip_ids.iter().any(|tip_id| {
+                if let Some(tip) = state.memory.tips.get(tip_id) {
+                    return tip.index == index && tip.canister_id == canister_id;
+                }
+                false
+            })
+        })
+    {
+        return Err("tip external index already exists".to_string());
+    }
+
+    if Some(true) != has_external_tip {
+        Post::mutate(state, &post_id, |post| {
+            post.has_external_tip = Some(true);
+            Ok(())
+        })
+        .expect("post not found");
+    }
+
+    let mut tip = Tip::new(post_id, sender_id, canister_id, amount, index);
+    tip.id = state.memory.tips.len() as u64 + 1;
+
+    // Insert to stable memory
+    state.memory.tips.insert(tip.id, tip.clone())?;
+
+    // Create indexes on heap
+    if let Some(tip_ids) = state.post_tip_indexes.get_mut(&post_id) {
+        tip_ids.push(tip.id);
+    } else {
+        state.post_tip_indexes.insert(post_id, vec![tip.id]);
+    }
+
+    state
+        .users
+        .get_mut(&receiver_id)
+        .expect("user not found")
+        .notify_about_post(
+            format!("@{} tipped you with for your post", sender_name),
+            post_id,
+        );
+
+    Ok(tip)
+}
+
+fn convert_memo_to_u64(memo: Option<Vec<u8>>) -> Result<PostId, String> {
+    let memo_value = memo.ok_or("memo is not defined")?;
+
+    let mut vec = memo_value.clone();
+
+    // Add leading zeros if needed
+    let current_len = vec.len();
+    if current_len < 8 {
+        let zeros_to_add = 8 - current_len;
+        vec.splice(0..0, vec![0; zeros_to_add]);
+    }
+
+    let mut num = 0u64;
+
+    // Combine bytes in little-endian order
+    for &byte in vec.iter() {
+        num = (num << 8) | byte as u64; // Shift left and add byte
+    }
+
+    Ok(num)
```

Comment:
I didn't test it, but this is more idiomatic, please TAL:
```suggestion
let mut padded = [0u8; 8];
let len = std::cmp::min(memo_value.len(), 8);
padded[8-len..].copy_from_slice(&memo_value[..len]);
Ok(u64::from_be_bytes(padded))
```

Reply:
amazing, works the same. How did you spot this ?